enhancements/network/egress-qos.md

```diff
@@ -0,0 +1,119 @@
+
+## Motivation
+
+Telco customers require support for DSCP marking capability for some of their 5G applications, giving some pods precedence over others.
+The QoS markings will be consumed and acted upon by objects outside of the OpenShift cluster to optimize traffic flow throughout their networks.
+
+### Goals
+
+- Provide a mechanism for users to set DSCP on egress traffic coming from specific namespaces.
+
+### Non-Goals
+
+- Ingress QoS.
+
+- Consolidating with current `kubernetes.io/egress-bandwidth` and `kubernetes.io/ingress-bandwidth` annotations.
+Nonetheless, the work done here does not interfere with the current bandwidth QoS mechanism.
+
+- The DSCP marking does not need to be handled or acted upon by OpenShift, just added to selected headers.
+
+- Marking East/West traffic, exposing the DSCP value from the inner packet to the outer geneve packet.
+
+## Proposal
+
+To achieve egress DSCP marking on pods, we introduce a new CRD `EgressQos`, which allows specifying a set of DSCP values and destination CIDRs - meaning traffic coming from pods on the namespace heading to each destination CIDR will be marked with the given DSCP value.
+
+```yaml
+kind: EgressQos
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: default
+  namespace: default
+spec:
+  egress:
+  - dscp: 46
+    dstCIDR: 0.0.0.0/0
+    priority: "?" # Do we add a priority field to handle overlapping CIDRs? Should we handle that in the code instead?
+  - dscp: 30
+    dstCIDR: 1.2.3.4/32
+```
+
+OVN-K watches these resources, creating/updating/deleting QoS objects in nbctl accordingly.
+For example, assuming there's a single pod `app` in namespace `default` on node `node1` and the above `EgressQos` created the equivalent of:
+```bash
+ovn-nbctl qos-add node1 from-lport 1 "ip4.src == <default_ns_address_set> && ip4.dst == 0.0.0.0/0" dscp=46
+ovn-nbctl qos-add node1 from-lport 2 "ip4.src == <default_ns_address_set> && ip4.dst == 1.2.3.4/32" dscp=30
+```
+will be executed.
+
+In addition it'll watch pods to decide if further updates are needed, for example attaching the QoS to another switch.
+
+### User Stories
+
+### API Extensions
+
+A new namespace-scoped CRD is introduced.
+### Implementation Details/Notes/Constraints
+
+- Adding a new API `EgressQos` under the `k8s.ovn.org/v1` version to `pkg/crd`.
+
+- Adding a controller that watches `EgressQoses`, `Pods`, producing the relevant QoS objects in nbdb and attaching them to the join switch or to the relavant logical switches in LGW.
+
+### Risks and Mitigations
+
+## Design Details
+
+### Test Plan
+
+* Unit tests coverage
+
+* IPv4/IPv6 E2E that validates egress traffic from a namespace is marked with the correct DSCP value by creating and deleting `EgressQoses`, setting up src pods and host-networked destination pods.
+  * Traffic to the specified CIDR should be marked.
+  * Traffic to an address not contained in the CIDR should not be marked.
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
```

Comment:
This is Dev Preview in 4.11 right?
Tech Preview in 4.12 and GA in 4.13? Is that the goal?